// @kbn/securitysolution-t-grid/target_node/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ROW_RENDERER_CLASS_NAME = exports.NOTE_CONTENT_CLASS_NAME = exports.NOTES_CONTAINER_CLASS_NAME = exports.KEYBOARD_DRAG_OFFSET = exports.IS_DRAGGING_CLASS_NAME = exports.HOVER_ACTIONS_ALWAYS_SHOW_CLASS_NAME = exports.HIGHLIGHTED_DROP_TARGET_CLASS_NAME = exports.EMPTY_PROVIDERS_GROUP_CLASS_NAME = exports.DRAGGABLE_KEYBOARD_WRAPPER_CLASS_NAME = void 0;

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
const HIGHLIGHTED_DROP_TARGET_CLASS_NAME = 'highlighted-drop-target';
exports.HIGHLIGHTED_DROP_TARGET_CLASS_NAME = HIGHLIGHTED_DROP_TARGET_CLASS_NAME;
const EMPTY_PROVIDERS_GROUP_CLASS_NAME = 'empty-providers-group';
/** The draggable will move this many pixels via the keyboard when the arrow key is pressed */

exports.EMPTY_PROVIDERS_GROUP_CLASS_NAME = EMPTY_PROVIDERS_GROUP_CLASS_NAME;
const KEYBOARD_DRAG_OFFSET = 20;
exports.KEYBOARD_DRAG_OFFSET = KEYBOARD_DRAG_OFFSET;
const DRAGGABLE_KEYBOARD_WRAPPER_CLASS_NAME = 'draggable-keyboard-wrapper';
exports.DRAGGABLE_KEYBOARD_WRAPPER_CLASS_NAME = DRAGGABLE_KEYBOARD_WRAPPER_CLASS_NAME;
const ROW_RENDERER_CLASS_NAME = 'row-renderer';
exports.ROW_RENDERER_CLASS_NAME = ROW_RENDERER_CLASS_NAME;
const NOTES_CONTAINER_CLASS_NAME = 'notes-container';
exports.NOTES_CONTAINER_CLASS_NAME = NOTES_CONTAINER_CLASS_NAME;
const NOTE_CONTENT_CLASS_NAME = 'note-content';
/** This class is added to the document body while dragging */

exports.NOTE_CONTENT_CLASS_NAME = NOTE_CONTENT_CLASS_NAME;
const IS_DRAGGING_CLASS_NAME = 'is-dragging';
exports.IS_DRAGGING_CLASS_NAME = IS_DRAGGING_CLASS_NAME;
const HOVER_ACTIONS_ALWAYS_SHOW_CLASS_NAME = 'hover-actions-always-show';
exports.HOVER_ACTIONS_ALWAYS_SHOW_CLASS_NAME = HOVER_ACTIONS_ALWAYS_SHOW_CLASS_NAME;
