// @elastic/eui/lib/components/datagrid/__mocks__/row_height_utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockRowHeightUtils = void 0;

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
var mockRowHeightUtils = {
  computeStylesForGridCell: jest.fn(),
  setGrid: jest.fn(),
  getStylesForCell: jest.fn(function () {
    return {
      wordWrap: 'break-word',
      wordBreak: 'break-word',
      flexGrow: 1
    };
  }),
  isDefinedHeight: jest.fn(function () {
    return true;
  }),
  isAutoHeight: jest.fn(function () {
    return false;
  }),
  setRowHeight: jest.fn(),
  pruneHiddenColumnHeights: jest.fn(),
  getRowHeight: jest.fn(function () {
    return 32;
  }),
  getComputedCellStyles: jest.fn(function () {}),
  getCalculatedHeight: jest.fn(function (heightOption, defaultHeight) {
    if (_typeof(heightOption) === 'object') {
      if (heightOption.lineCount) {
        return heightOption.lineCount;
      }

      if (heightOption.height) {
        return heightOption.height;
      }
    }

    if (heightOption) {
      return heightOption;
    }

    return defaultHeight;
  })
};
exports.mockRowHeightUtils = mockRowHeightUtils;
