// @elastic/eui/lib/components/datagrid/data_grid_context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DataGridWrapperRowsContext = exports.DataGridSortingContext = exports.DataGridFocusContext = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
var DataGridFocusContext = /*#__PURE__*/_react.default.createContext({
  setFocusedCell: function setFocusedCell() {},
  onFocusUpdate: function onFocusUpdate() {
    return function () {};
  }
});

exports.DataGridFocusContext = DataGridFocusContext;

var DataGridSortingContext = /*#__PURE__*/_react.default.createContext(undefined);

exports.DataGridSortingContext = DataGridSortingContext;

var DataGridWrapperRowsContext = /*#__PURE__*/_react.default.createContext({
  headerRow: /*#__PURE__*/_react.default.createElement("div", null),
  headerRowHeight: 0,
  footerRow: null
});

exports.DataGridWrapperRowsContext = DataGridWrapperRowsContext;
